packages/appium/lib/extension/extension-config.js

```diff
@@ -0,0 +1,458 @@
+  INSTALL_TYPE_GITHUB,
+  INSTALL_TYPE_LOCAL,
+  INSTALL_TYPE_NPM,
+]);
+
+/**
+ * This class is abstract. It should not be instantiated directly.
+ *
+ * Subclasses should provide the generic parameter to implement.
+ * @template {ExtensionType} ExtType
+ */
+export class ExtensionConfig {
+  /** @type {ExtType} */
+  extensionType;
+
+  /** @type {`${ExtType}s`} */
+  configKey;
+
+  /** @type {ExtRecord<ExtType>} */
+  installedExtensions;
+
+  /** @type {ExtensionLogFn} */
+  log;
+
+  /** @type {import('./manifest').Manifest} */
+  io;
+
+  /**
+   * @protected
+   * @param {ExtType} extensionType - Type of extension
+   * @param {import('./manifest').Manifest} io - IO object
+   * @param {ExtensionLogFn} [logFn]
+   */
+  constructor (extensionType, io, logFn) {
+    const logger = _.isFunction(logFn) ? logFn : log.error.bind(log);
+    this.extensionType = extensionType;
+    this.configKey = `${extensionType}s`;
+    this.installedExtensions = io.getExtensionData(extensionType);
+    this.log = logger;
+    this.io = io;
+  }
+
+  get manifestPath () {
+    return this.io.manifestPath;
+  }
+
+  get appiumHome () {
+    return this.io.appiumHome;
+  }
+
+  /**
+   * Checks extensions for problems
+   * @param {ExtRecord<ExtType>} exts - Extension data
+   */
+  validate (exts) {
+    const foundProblems =
+      /** @type {Record<ExtName<ExtType>,Problem[]>} */ ({});
+    for (const [
+      extName,
+      extData,
+    ] of /** @type {[ExtName<ExtType>, ExtData<ExtType>][]} */ (
+        _.toPairs(exts)
+      )) {
+      foundProblems[extName] = [
+        ...this.getGenericConfigProblems(extData, extName),
+        ...this.getConfigProblems(extData),
+        ...this.getSchemaProblems(extData, extName),
+      ];
+    }
+
+    const problemSummaries = [];
+    for (const [extName, problems] of _.toPairs(foundProblems)) {
+      if (_.isEmpty(problems)) {
+        continue;
+      }
+      // remove this extension from the list since it's not valid
+      delete exts[extName];
+      problemSummaries.push(
+        `${this.extensionType} ${extName} had errors and will not ` +
+          `be available. Errors:`,
+      );
+      for (const problem of problems) {
+        problemSummaries.push(
+          `  - ${problem.err} (Actual value: ` +
+            `${JSON.stringify(problem.val)})`,
+        );
+      }
+    }
+
+    if (!_.isEmpty(problemSummaries)) {
+      this.log(
+        `Appium encountered one or more errors while validating ` +
+          `the ${this.configKey} extension file (${this.manifestPath}):`,
+      );
+      for (const summary of problemSummaries) {
+        this.log(summary);
+      }
+    }
+
+    return exts;
+  }
+
+  /**
+   * @param {ExtData<ExtType>} extData
+   * @param {ExtName<ExtType>} extName
+   * @returns {Problem[]}
+   */
+  getSchemaProblems (extData, extName) {
+    const problems = [];
+    const {schema: argSchemaPath} = extData;
+    if (ExtensionConfig.extDataHasSchema(extData)) {
```

Comment:
I prefer to keep such methods on module level and only make them static if we need this data in subclasses

Reply:
noted